diff -Naurp retroarch-ce9439e/command.c retroarch-ce9439e-patch/command.c
--- retroarch-ce9439e/command.c	2017-12-21 10:19:41.000000000 +0200
+++ retroarch-ce9439e-patch/command.c	2017-12-21 10:28:22.000000000 +0200
@@ -1974,8 +1974,8 @@ bool command_event(enum event_command cm
          break;
       case CMD_EVENT_QUIT:
          remove("/var/lock/start.retro");
-         creat("/var/lock/start.games", S_IRUSR|S_IROTH);
-         system("systemctl start emustation.service");
+         creat("/var/lock/start.kodi", S_IRUSR|S_IROTH);
+         system("systemctl start kodi.service");
          return retroarch_main_quit();
       case CMD_EVENT_CHEEVOS_HARDCORE_MODE_TOGGLE:
 #ifdef HAVE_CHEEVOS
@@ -2288,20 +2288,16 @@ bool command_event(enum event_command cm
          }
          break;
       case CMD_EVENT_SHUTDOWN:
-#if defined(__linux__) && !defined(ANDROID)
          runloop_msg_queue_push(msg_hash_to_str(MSG_VALUE_SHUTTING_DOWN), 1, 180, true);
          command_event(CMD_EVENT_MENU_SAVE_CURRENT_CONFIG, NULL);
          command_event(CMD_EVENT_QUIT, NULL);
-         system("shutdown -P now");
-#endif
+         system("systemctl poweroff");
          break;
       case CMD_EVENT_REBOOT:
-#if defined(__linux__) && !defined(ANDROID)
          runloop_msg_queue_push(msg_hash_to_str(MSG_VALUE_REBOOTING), 1, 180, true);
          command_event(CMD_EVENT_MENU_SAVE_CURRENT_CONFIG, NULL);
          command_event(CMD_EVENT_QUIT, NULL);
-         system("shutdown -r now");
-#endif
+         system("systemctl reboot");
          break;
       case CMD_EVENT_RESUME:
          rarch_ctl(RARCH_CTL_MENU_RUNNING_FINISHED, NULL);
