From 6b154d3011a07019210b1b8d0e96fb03c23b97ba Mon Sep 17 00:00:00 2001
From: Sam Nazarko <email@samnazarko.co.uk>
Date: Mon, 2 Apr 2018 02:29:34 +0100
Subject: [PATCH] audiodsp: reset audio chip when leaving passthrough

---
 drivers/amlogic/audiodsp/audiodsp_module.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/amlogic/audiodsp/audiodsp_module.c b/drivers/amlogic/audiodsp/audiodsp_module.c
index 882ffa3f64d..34e5cd973c6 100644
--- a/drivers/amlogic/audiodsp/audiodsp_module.c
+++ b/drivers/amlogic/audiodsp/audiodsp_module.c
@@ -825,6 +825,14 @@ static ssize_t digital_codec_store(struct class *class,
 	 * need trigger pcm hw prepare to re-init hw configuration
 	 */
 	pr_info("last mode %d,now %d\n", mode_codec, IEC958_mode_codec);
+	/* reset audio chip for pcm if necessary */
+	if (IEC958_mode_codec == 0) {
+	    pr_info("putting chip back to PCM state");
+	    aml_alsa_hw_reprepare();
+	    struct audiodsp_priv *priv = audiodsp_privdata();
+            priv->decoded_nb_frames = 0;
+            priv->format_wait_count = 0;
+	}
 
 	return count;
 }
