diff -Naurp kodi-17.5-rc1-07424d3/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.h kodi-17.5-rc1-07424d3-patch/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.h
--- kodi-17.5-rc1-07424d3/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.h	2017-09-20 08:21:44.000000000 +0300
+++ kodi-17.5-rc1-07424d3-patch/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.h	2017-09-20 09:29:41.000000000 +0300
@@ -62,7 +62,7 @@ public:
 private:
 
   int m_iRenderBuffer;
-  static const int m_numRenderBuffers = 4;
+  static const int m_numRenderBuffers = 6;
 
   struct BUFFER
   {
